### ISSUE TITLE
Dump firmware

### ISSUE THREAD

**Issue by suhajda3** [v0l/radio_tool#15]
Hello,

Is it possible to dump the whole firmware? I have two UV-15R and accidentally bricked one of them with a wrong firmware. I would like to download the original firmware from the good one and upload it to the bricked one.

Thanks!

**Comment by v0l** [comment]
Are you able to start the radio in programming mode?

**Comment by suhajda3** [comment]
Yes, I can start programming mode on both radios.

**Comment by v0l** [comment]
I think you can just flash the firmware on again by downloading it online? There shouldnt be a need to copy the firmware from one radio to the other

**Comment by suhajda3** [comment]
I couldn't find the firmware online, that is why I would like to dump it from the good one.

**Comment by v0l** [comment]
Ok, there is a dump command but it only works on Mac, do you have a macbook?

**Comment by v0l** [comment]
There is no support currently to dump the firmware only the bootloader

**Comment by suhajda3** [comment]
Yes, I can run it on a Mac. Do you have any plans to support firmware dump as well?

**Comment by v0l** [comment]
I dont have a lot of spare time but i can try to see if it works in the next few days, ill update here if it worked

**Comment by v0l** [comment]
I dont recall does this look like vector table? 
![image](https://user-images.githubusercontent.com/1172179/185504599-369906d1-5e1e-4c20-bbf5-e54da3359960.png)

**Comment by v0l** [comment]
Ok it looks like it does just dump the decrypted firmware.. need to finish this off

**Comment by kosciej** [comment]
What's needed to finish this issue? Maybe I can help with it. I have exactly the same problem as @suhajda3 - I've installed P15UV GMRS firmware on UV-15R radio and want to go back.

**Comment by v0l** [comment]
Test this branch and see if works for your radio, it might not even work as some bootloaders dont allow reading at all.

https://github.com/v0l/radio_tool/tree/try-dump-firmware

**Comment by suhajda3** [comment]
Thanks.

I get this error when running the command:

`error parsing options: Argument ‘-o’ failed to parse`

**Comment by v0l** [comment]
Try with ``--dump-firmware` at the end

**Comment by suhajda3** [comment]
./radio_tool -d 0 -o test.bin --dump-firmware
`error parsing options: Option ‘dump-firmware’ is missing an argument`

**Comment by v0l** [comment]
Ok add a random value `--dump-firmware true`

**Comment by suhajda3** [comment]
./radio_tool -d 0 -o test.bin --dump-firmware true
`error parsing options: Argument ‘true’ failed to parse`

**Comment by v0l** [comment]
Looks like its supposed to be a number, try with a number

**Comment by suhajda3** [comment]
./radio_tool -d 0 -o test.bin --dump-firmware 1
`Segmentation fault: 11`

**Comment by v0l** [comment]
Maybe try a larger number `0x1000`

**Comment by v0l** [comment]
Join [Discord](https://discord.gg/V596cyatmv) to discuss more

**Comment by drzraf** [comment]
Cross-referencing https://github.com/CtrlC-Root/gm30/issues/1 in case it could help.
